synapse/config/_util.py

```diff
@@ -11,10 +11,16 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-from typing import Any, Dict, Type, TypeVar
+from typing import TYPE_CHECKING, Any, Dict, Type, TypeVar
 
 import jsonschema
-from pydantic import BaseModel, ValidationError, parse_obj_as
+
+from synapse._pydantic_compat import HAS_PYDANTIC_V2
+
+if TYPE_CHECKING or HAS_PYDANTIC_V2:
+    from pydantic.v1 import BaseModel, ValidationError, parse_obj_as
+else:
+    from pydantic import BaseModel, ValidationError, parse_obj_as
```

Comment:
Why not a try-except on the import error?

Reply:
This doesn't work with mypy. It doesn't know that you except `pydantic.v1` when pydantic v2 is installed and plain `pydantic` when pydantic v1 is installed and reports many spurious errors. Using `if TYPE_CHECKING:` makes it so mypy only follows the first import branch (`from pydantic.v1 import ...`) and ignores the second. Since the `pydantic.v1` module is only available with pydantic v2, I pinned it in the dev-dependencies.

If you really want to get rid of `HAS_PYDANTIC_V2`, something like this would have the same effect, but is incredibly ugly.

```suggestion
if TYPE_CHECKING:
    from pydantic.v1 import BaseModel, ValidationError, parse_obj_as
else:
    try:
        from pydantic.v1 import BaseModel, ValidationError, parse_obj_as
    except ImportError:
        from pydantic import BaseModel, ValidationError, parse_obj_as
```

Reply:
Thanks for the explanation! I agree that's incredibly ugly. 👍